### ISSUE TITLE
LoadFromFile with InPlaceResourceOptimization problems.

### ISSUE THREAD

**Issue by tekert** [pagespeed/ngx_pagespeed#1252]
This only happens when
```
pagespeed InPlaceResourceOptimization on; (its on by default now)
pagespeed LoadFromFile LOCAL_PATH; (defined)
```

If `pagespeed InPlaceResourceOptimization off`, these errors won't happen.
Noticed this while hunting for info on why `Accept-Ranges: bytes` were removed on some requests.

Problems:
1.- Already optimized resources (not by pagespeed) are missing important response headers.
2.- Server wont honor If-Modified-Since request header on valid resource.

PageSpeed config:
```
pagespeed RewriteLevel CoreFilters;
pagespeed ImplicitCacheTtlMs 864000000;
pagespeed LoadFromFile VALID_PATH;
pagespeed EnableFilters in_place_optimize_for_browser;
```

These test where made with an image, also happens with some js and css files.

Optimized JPG
--------------------
_(Without LoadFromFile this request is the same as PageSpeed=off)_
```
curl -I https://mydomain.com/image1_optimized.jpg?PageSpeed=on
HTTP/2 200
server: nginx
content-type: image/jpeg
date: Fri, 19 Aug 2016 01:43:39 GMT
expires: Mon, 29 Aug 2016 01:43:39 GMT
cache-control: max-age=864000
last-modified: Fri, 19 Aug 2016 01:05:24 GMT
```

_Missing etag, content-length, accept-ranges_

```
curl -I https://mydomain.com/image1_optimized.jpg?PageSpeed=off
HTTP/2 200
server: nginx
date: Fri, 19 Aug 2016 01:43:22 GMT
content-type: image/jpeg
content-length: 11922
last-modified: Fri, 19 Aug 2016 01:05:24 GMT
etag: "57b65b54-2e92"
accept-ranges: bytes
```

Same but unoptimized JPG
-------

```
curl -I https://mydomain.com/image1_UNoptimized.jpg?PageSpeed=on
HTTP/2 200
server: nginx
content-type: image/jpeg
content-length: 14572
etag: W/"PSA-aj-87pMRAf8_b"
date: Fri, 19 Aug 2016 03:43:50 GMT
expires: Mon, 29 Aug 2016 03:43:50 GMT
cache-control: max-age=864000
vary: User-Agent,Save-Data
```

_seems normal except again accept-ranges: bytes is missing._

```
curl -I https://mydomain.com/image1_UNoptimized.jpg?PageSpeed=off
HTTP/2 200
server: nginx
date: Fri, 19 Aug 2016 03:44:12 GMT
content-type: image/jpeg
content-length: 25924
last-modified: Fri, 19 Aug 2016 01:08:23 GMT
etag: "57b65c07-6544"
accept-ranges: bytes
```

Optimized JPG (Same test but with LoadFromFile not defined)
-----------------------------------------------------------------

```
curl -I https://mydomain.com/image1_optimized.jpg?PageSpeed=on
HTTP/2 200
server: nginx
date: Fri, 19 Aug 2016 01:43:22 GMT
content-type: image/jpeg
content-length: 11922
last-modified: Fri, 19 Aug 2016 01:05:24 GMT
etag: "57b65b54-2e92"
accept-ranges: bytes
```

Same response as PageSpeed=off (yes, tested with multiple GETs over time)

Conditional cache request not wotking
-----------------------------------------------------------------
### Now, if for example we do this with LoadFromFile defined:

```
curl -I https://mydomain.com/image1_optimized.jpg?PageSpeed=on -H "If-Modified-Since: Fri, 19 Aug 2016 01:05:24 GMT"
HTTP/2 200
server: nginx
content-type: image/jpeg
date: Fri, 19 Aug 2016 09:36:57 GMT
expires: Mon, 29 Aug 2016 09:36:57 GMT
cache-control: max-age=864000
last-modified: Fri, 19 Aug 2016 01:05:24 GMT
```

_this should be HTTP/2 304 ..._

### Same test with LoadFromFile NOT defined:

```
curl -I https://mydomain.com/image1_optimized.jpg?PageSpeed=on -H "If-Modified-Since: Fri, 19 Aug 2016 01:05:24 GMT" -H "If-None-Match: \"57b65b54-2e92\""
HTTP/1.1 200 OK
Server: nginx
Date: Sat, 20 Aug 2016 03:50:45 GMT
Content-Type: image/jpeg
Content-Length: 11922
Last-Modified: Fri, 19 Aug 2016 01:05:24 GMT
Connection: keep-alive
ETag: "57b65b54-2e92"
Accept-Ranges: bytes
```

_this should be HTTP/2 304 ..._
_NOTE: exactly after 5min, it will return 304._

Same thing happens while requesting an html page with those resources on valid browser headers.
Same thing happens when using HTTP/1.1.

Wonder where is the problem.

pagespeed ver. 1.11.33.2-0

**Comment by huibaolin** [comment]
@tekert Thanks for the detailed analysis!

I think we've 2 issues: (1) behavior of LoadFromFile and (2) "accept-range" header (covered in other thread).

**Comment by tekert** [comment]
Yes, it seems there may be two separate problems.
Concerning problem with *2:

Post this in case someone wants to reproduce this in a browser:

1) Make an html and put some resources, some already optimized, some not, etc. Or use any html with many resources.
2) Request that html with PageSpeed=off (to get the original unmodified html)
3) Refresh the page.
4) Watch how some resources don't get code 304 on valid conditional request.

**Comment by jeffkaufman** [comment]
Very thorough bug report, thanks!

> Optimized JPG: missing etag, content-length, accept-ranges

With LoadFromFile we skip the host webserver's serving path, and load directly from disk.

I've filed a bug (#1381) for serving it as chunked instead of giving a content length, and for etag not being present (#1382).  Range isn't going to be included since we don't support range requests.

> Conditional cache request not working

In general, we support etag-based conditional refreshes, and not last-modified ones.  When we fix etag not being present (#1382) we'll remove the last-modified header to make sure browsers only send us if-not-modified.

**Comment by jeffkaufman** [comment]
Closing this bug, since it's been repro'd and broken down into #1381 and #1382.

If there are other issues I'm missing that haven't been captured by those two focused cases, feel free to reopen.
